### ISSUE TITLE
On the issue of tag structure

### ISSUE THREAD

**Issue by pine-fish** [littlefs-project/littlefs#1000]
I tried to dump the LFS data and read it, but I found that the structure introduced in SPEC.md does not match my data
Here is a portion of my data
`02 00 00 00 F0 0F FF F7 6C 69 74 74 6C 65 66 73 2F E0 00 10 00 00 02 00 00 10 00 00 8E 00 00 00 FF 00 00 00 FF FF FF 7F FE 03 00 00 70 1F FC 08 F0 28 5E BC FF FF FF FF FF FF FF FF FF FF FF FF 10 1F F8 10 40 00 00 07 31 31 31 2E 62 69 6E 20 00 00 07 70 1F F8 09 DA CE D0 DD FF FF FF FF FF 10 1F F4 09 40 00 00 08 32 32 32 32 2E 68 65 78 20 00 00 08 70 1F F4 08 DE E3 63 56 FF FF FF FF 10 1F F0 08 40 00 00 09 33 33 33 33 33 2E 6A 70 67 20 00 00 09 70 1F F0 07 01 C5 9C E7 FF FF FF`

![1](https://github.com/littlefs-project/littlefs/assets/174169932/8259e63c-fe30-440a-8a2c-d8f2e1d442b9)
I found the file names 111.bin, 2222hex, and 33333.jpg, but they have not saved any data yet. 
Taking 111. bin as an example, the first few bytes of the file name should be its corresponding tag, right? The first byte 07 before the file name also corresponds to the length of the file name, but a few more bytes before it seem to have a different file structure than the one in the screenshot? I cannot read any correct information based on this structure. 
Is it because the tag was XOR that the data cannot be read directly? Can anyone explain this structure in layman's terms? Thanks!

**Comment by geky** [comment]
Hi @pine-fish, sorry about the late response

The tag encoding described in [SPEC.md](https://github.com/littlefs-project/littlefs/blob/d01280e64934a09ba16cac60cf9d3a37e228bb66/SPEC.md) is only relevant after decoding the xor structure, which is a bit annoying to do by hand.

The xor starts with 0xffffffff, so for the first tag (skipping the revision count):

```
    F0 0F FF F7
xor ff ff ff ff
 => 0f f0 00 08
```

We can then see the tag structure:

```
   0f f0 00 08
   '--|---|--|--> valid           (tag[31]=0x0)
      '---|--|--> type=superblock (tag[20:31]=0x0ff)
          '--|--> id=0            (tag[10:20]=0x000)
             '--> size=8          (tag[0:10]=0x0ff)
```

To decode the next tag, we xor with the most recent tag, so for the second tag:

```
    2F E0 00 10
xor 0f f0 00 08
 => 20 10 00 18
```

And so on.

You can probably see where this gets a bit tedious, so I usually use the [readmdir.py](https://github.com/littlefs-project/littlefs/blob/d01280e64934a09ba16cac60cf9d3a37e228bb66/scripts/readmdir.py) script in the scripts dir:

``` bash
$ echo "02 00 00 00 F0 0F FF F7 6C 69 74 74 6C 65 66 73 2F E0 00 10 00 00 02 00 00 10 00 00 8E 00 00 00 FF 00 00 00 FF FF FF 7F FE 03 00 00 70 1F FC 08 F0 28 5E BC FF FF FF FF FF FF FF FF FF FF FF FF 10 1F F8 10 40 00 00 07 31 31 31 2E 62 69 6E 20 00 00 07 70 1F F8 09 DA CE D0 DD FF FF FF FF FF 10 1F F4 09 40 00 00 08 32 32 32 32 2E 68 65 78 20 00 00 08 70 1F F4 08 DE E3 63 56 FF FF FF FF 10 1F F0 08 40 00 00 09 33 33 33 33 33 2E 6A 70 67 20 00 00 09 70 1F F0 07 01 C5 9C E7 FF FF FF" | xxd -r -p > disk
$ ./scripts/readmdir.py disk 4096 0
mdir {0x0} rev 2
off       tag       type            id  len  data (truncated)
00000008: 0ff00008  superblock       0    8  6c 69 74 74 6c 65 66 73  littlefs
00000014: 20100018  inlinestruct     0   24  00 00 02 00 00 10 00 00  ........
00000048: 00100407  reg              1    7  31 31 31 2e 62 69 6e     111.bin
00000053: 20100400  inlinestruct     1    0
00000068: 00100808  reg              2    8  32 32 32 32 2e 68 65 78  2222.hex
00000074: 20100800  inlinestruct     2    0
00000088: 00100c09  reg              3    9  33 33 33 33 33 2e 6a 70  33333.jp
00000095: 20100c00  inlinestruct     3    0
```

Though note this script is a bit hacky and just provided as is.

The reason for this xor stuff is it also works in reverse if you know the last tag, allowing iteration backwards through the log. Though a simpler design may have just used two tags, one before and after, for each piece of metadata.
